src/hotspot/share/opto/parseHelper.cpp

```diff
@@ -869,9 +865,245 @@ Node* PEAState::get_java_oop(ObjID id) const {
+      }
+      return retval;
+  };
+
+  if (shim == nullptr) {
+      shim = new RegionNode(2);
+      shim->set_req(1, r->in(pnum));
+      kit->gvn().set_type(shim, Type::CONTROL);
+      kit->record_for_igvn(shim);
+
+      map = newin_map->clone()->as_SafePoint();
+      map->set_req(0, shim); // attach to shim!
+      retval = mat(shim, map, var, r, pnum);
+  } else {
+    assert(map != nullptr, "must associate with a SafePoint");
+    shim->add_req(r->in(pnum));
+    ObjectState* os = get_object_state(id);
+    if (os->is_virtual()) {
+        retval = mat(shim, map, var, r, pnum);
+    } else {
+        retval = phi->in(path);
+    }
+  }
+
+  phi->set_req(pnum, retval);
+  return retval;
+}
+
+// materialize var for all predecessors of current region.
+// r->in(pnum) is newin and the object is escaped in newin.
+// We need to clone kit->map() and rewrite those phi nodes for the predecessors.
+// after materialization, the new code controls all covered paths.
+Node* PEAState::materialize_incoming_escaped(GraphKit* kit, Node* var, RegionNode* r, uint pnum) {
+  if (!EnablePassiveMaterialization) return var;
+
+  auto* pea = kit->PEA();
+  ObjID id = pea->is_alias(var);
+  PhaseGVN& gvn = kit->gvn();
+
+  RegionNode* shim = new RegionNode(1);
+  GrowableArray<uint> paths;
+
+  for (uint i=1; i<r->req(); ++i) {
+    if (i == pnum) continue;
+    Node* n = r->in(i);
+    if (n != nullptr) {
+      paths.append(i);
+      shim->add_req(n);
+    }
+  }
+  gvn.set_type(shim, Type::CONTROL);
+  kit->record_for_igvn(shim);
+
+  SafePointNode* map = kit->map()->clone()->as_SafePoint();
+  map->set_req(0, shim); // attach to shim!
+  GraphKit shim_kit(map->jvms());
+  bool unique = paths.length() <= 1;
+
+  // shim only covers selective paths, so we have to rewrite inputs if they are phi nodes.
+  for (uint i = 1; i < map->req(); ++i) {
+    Node* n = map->in(i);
+
+    if (i == TypeFunc::Memory && n->is_MergeMem()) {
+      // a blank memory
+      MergeMemNode* new_mem = MergeMemNode::make(MergeMemNode::make_empty_memory());
+      new_mem->grow_to_match(n->as_MergeMem());
+      for (MergeMemStream mms(n->as_MergeMem()); mms.next_non_empty(); ) {
+        Node* p = mms.memory();
+
+        if (p->is_Phi() && p->in(0) == r) {
+          if (unique) {
+            new_mem->set_req(mms.alias_idx(), p->in(paths.at(0)));
+          } else {
+            PhiNode* phi = PhiNode::make(shim, nullptr, Type::MEMORY, p->adr_type());
+            for (int j =0; j < paths.length(); ++j) {
+              int idx = paths.at(j);
+              phi->set_req(j + 1, p->in(idx));
+            }
+            gvn.set_type(phi, Type::MEMORY);
+            new_mem->set_req(mms.alias_idx(), kit->gvn().transform(phi));
+          }
+        } else {
+          new_mem->set_req(mms.alias_idx(), p);
+        }
+      }
+      map->set_req(i, new_mem);
+    } else if (n->is_Phi() && n->in(0) == r) {
+      if (unique) {
+        map->set_req(i, n->in(paths.at(0)));
```

Comment:
Would it make sense to make this a phi with a single input? Because if you merge in a virtual object that needs to be passively materialized, you will use the same shim and will need to create a phi at that point.